drivers/spi/spi_gecko_eusart.c

```diff
@@ -0,0 +1,309 @@
+
+	/* Initialize the EUSART */
+	EUSART_SpiInit(gecko_config->base, &eusartInit);
+
+	data->ctx.config = config;
+
+	/* Enable the peripheral */
+	gecko_config->base->CMD = (uint32_t)eusartEnable;
+
+	return 0;
+}
+
+static void spi_gecko_eusart_send(EUSART_TypeDef *eusart, uint8_t frame)
+{
+	/* Write frame to register */
+	EUSART_Tx(eusart, frame);
+
+	/* Wait until the transfer ends */
+	while (!(eusart->STATUS & EUSART_STATUS_TXC)) {
+	}
+}
+
+static uint8_t spi_gecko_eusart_recv(EUSART_TypeDef *eusart)
+{
+	/* Return data inside rx register */
+	return EUSART_Rx(eusart);
+}
+
+static bool spi_eusart_transfer_ongoing(struct spi_gecko_eusart_data *data)
+{
+	return spi_context_tx_on(&data->ctx) || spi_context_rx_on(&data->ctx);
+}
+
+static inline uint8_t spi_eusart_next_tx(struct spi_gecko_eusart_data *data)
+{
+	uint8_t tx_frame = 0;
+
+	if (spi_context_tx_buf_on(&data->ctx)) {
+		tx_frame = UNALIGNED_GET((uint8_t *)(data->ctx.tx_buf));
+	}
+
+	return tx_frame;
+}
+
+static int spi_eusart_shift_frames(EUSART_TypeDef *eusart, struct spi_gecko_eusart_data *data)
+{
+	uint8_t tx_frame;
+	uint8_t rx_frame;
+
+	tx_frame = spi_eusart_next_tx(data);
+	spi_gecko_eusart_send(eusart, tx_frame);
+	spi_context_update_tx(&data->ctx, 1, 1);
+
+	rx_frame = spi_gecko_eusart_recv(eusart);
+
+	if (spi_context_rx_buf_on(&data->ctx)) {
+		UNALIGNED_PUT(rx_frame, (uint8_t *)data->ctx.rx_buf);
+	}
+	spi_context_update_rx(&data->ctx, 1, 1);
+	return 0;
+}
+
+static void spi_gecko_eusart_xfer(const struct device *dev, const struct spi_config *config)
+{
+	int ret;
+	struct spi_gecko_eusart_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	const struct spi_gecko_eusart_config *gecko_config = dev->config;
+
+	spi_context_cs_control(ctx, true);
+
+	do {
+		ret = spi_eusart_shift_frames(gecko_config->base, data);
+	} while (!ret && spi_eusart_transfer_ongoing(data));
+
+	spi_context_cs_control(ctx, false);
+	spi_context_complete(ctx, dev, 0);
+}
+
+/* API Functions */
+static int spi_gecko_eusart_init(const struct device *dev)
+{
+	int err;
+	const struct spi_gecko_eusart_config *config = dev->config;
+	struct spi_gecko_eusart_data *data = dev->data;
+
+	err = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (err < 0) {
+		return err;
+	}
+
+	err = spi_context_cs_configure_all(&data->ctx);
+	if (err < 0) {
+		return err;
+	}
+	spi_context_unlock_unconditionally(&data->ctx);
+
+	return 0;
+}
+
+static int spi_gecko_eusart_transceive(const struct device *dev, const struct spi_config *config,
+				       const struct spi_buf_set *tx_bufs,
+				       const struct spi_buf_set *rx_bufs)
+{
+	struct spi_gecko_eusart_data *data = dev->data;
+	uint16_t control = 0;
+	int ret;
+
+	spi_context_lock(&data->ctx, false, NULL, NULL, config);
+
+	ret = spi_eusart_config(dev, config, &control);
+	if (ret < 0) {
+		return ret;
+	}
+
+	spi_context_buffers_setup(&data->ctx, tx_bufs, rx_bufs, 1);
+	spi_gecko_eusart_xfer(dev, config);
+
```

Comment:
Hi guys,
Not quite sure this is relevant here, but we've been trying this new driver to circumvent some issues with the CAN MCP251xfd driver (currently randomly crashes in the eusart layer).
In our test of this SPI over EUSART, the driver is failing when we do multiple transceive (in a retry mechanism like the one in the read_crc from the MCP - https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/can/can_mcp251xfd.c#L83), due to both the spi_context lock semaphore never released (in transceive). I think adding a release of the lock is required here?

NB: we used the Silabs spi eusart driver as an out-of-tree driver until now and it was (mostly) working. Any hint

Reply:
spi_context_complete() does release the lock in spi_gecko_eusart_xfer()

Reply:
As per spi_context.h, not in all cases. It may only release the sync semaphore, not the lock

Reply:
Hello @Sylphe88 
I added the release in the transceive function, thanks for your comment!